aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1644,6 +1644,144 @@ Tensor& dstack_out(TensorList tensors, Tensor& result) {
+    const int64_t unsqu_size = result_dim - tensor.ndimension();
+    LowDimIdx unsqu_sizes;
+    for(auto j: c10::irange(unsqu_size)) {
+      unsqu_sizes.push_back(1);
+    }
+    for(auto j: c10::irange(unsqu_size, result_dim)) {
+      unsqu_sizes.push_back(tensor.size(j - unsqu_size));
+    }
+    tensor = tensor.view(unsqu_sizes);
+    ten_shapes.push_back(unsqu_sizes);
+
+    auto ten_ndim = tensor.ndimension();
+    auto idx = padded_idx[i];
+    auto next_idx = padded_idx[std::min(i + 1, padded_idx.size() - 1)];
+    LowDimIdx slice_a;
+    LowDimIdx slice_b;
+
+    // compute the slice in the result tensor that tensor_i will go, and
+    // update the cursor to the next position
+    for (auto j : c10::irange(result_dim)) {
+      slice_a.push_back(cursor[j]);
+
+      if (result_dim - j > ten_ndim)
+        slice_b.push_back(cursor[j] + 1);
+      else
+        slice_b.push_back(cursor[j] + tensor.size(j - (result_dim - ten_ndim)));
+      result_size[j] = (std::max(result_size[j], slice_b[j]));
+
+      if (next_idx[j] - idx[j] < 0)
+        cursor[j] = 0;
+      else
+        cursor[j] = cursor[j] + tensor.size(j) * (next_idx[j] - idx[j]);
+    }
+    slices_a.push_back(slice_a);
+    slices_b.push_back(slice_b);
+
+    // check is each "row" of tensors is concatable at dimension "j"
+    // and store the resulting tensor shape for next check at dimension "j + 1"
+    for(int64_t j = result_dim - 1; j >= 0; j--) {
+      const bool single_ele_carry = next_idx[j] - idx[j] == 0 &&
+          next_idx[std::max((int64_t)0, j - 1)] - idx[std::max((int64_t)0, j - 1)] > 0;
+      const bool id_carry = next_idx[j] - idx[j] < 0 || single_ele_carry;
+      const bool last_ten = i == tensors.size() - 1;
+
+      if (id_carry || last_ten) {
+        if (j == result_dim - 1) {
+          auto itr = ten_shapes.end() - (idx[j] + 1);
+          auto cated_shape = LowDimIdx(*itr);
+
+          for(auto a = itr + 1; a < ten_shapes.end(); a++) {
+            cated_shape = cat_shape(cated_shape, *a, j, padded_idx[i]);
+          }
+          hierar_shapes[j].push_back(cated_shape);
+        } else {
+          auto itr = hierar_shapes[j + 1].end() - (idx[j] + 1);
+          auto cated_shape = LowDimIdx(*itr);
+
+          for (auto a = itr + 1; a < hierar_shapes[j + 1].end(); a++) {
+            cated_shape = cat_shape(cated_shape, *a, j, padded_idx[i]);
+          }
+          hierar_shapes[j].push_back(cated_shape);
+        }
+      }
+    }
+
+  }
+
+  ScalarType output_scalar_type = native::result_type(tensors);
+  Tensor result = at::zeros(
+      result_size,
+      tensors[0].options().dtype(output_scalar_type));
+
+  for (auto i : c10::irange(tensors.size())) {
+    auto slice_a = slices_a[i];
+    auto slice_b = slices_b[i];
+    Tensor tensor = tensors[i];
+    Tensor ten_slic = result;
+    for (auto j: c10::irange(result_dim))
+      ten_slic = ten_slic.slice(j, slice_a[j], slice_b[j]);
+    ten_slic.copy_(tensor);
```

Comment:
This looks a bit questionable, in particular it doesn't seem like autograd is going to work with an implementation like this.

Reply:
You are right on that, after some testing, it does show that autograd's won't able to propagate gradient through **slices**. I think creating a block_backward to deal with the gradient will be the most reasonable solution for this?